### PR TITLE
Add the new error reporting api

### PR DESCRIPTION
**Pull request by diml**
This PR implements the new error reporting API described in #2190 and moves the old functions to the `Errors` module in the `dune` library. This means that stdune and all the libraries that come before `dune` have been updated to use the new API. In addition to what is described in #2190, I introduced a module `User_message` to represent all messages displayed to the user: error messages, warnings and informative messages. I also added an API for printing these on the terminal.

At the moment, `Report_error` was very roughly updated to handle the new exception. ~which explains why in one of the backbox test a spurious empty line is inserted. This will be sorted once more of the code has been ported to the `Pp` API. I intend to do this next.~ EDIT: I found why the empty line was inserted and fixed the issue.

### CODE REVIEW

File: src/stdune/pp.ml [hunk 1]
```diff
@@ -186,3 +199,8 @@ let text s = Text s
 let textf fmt = Printf.ksprintf text fmt
 
 let tag t ~tag = Tag (tag, t)
+
+let enumerate l ~f =
+  vbox (concat ~sep:cut (List.map l ~f:(fun x ->
+    box ~indent:2
+      (seq (char '-') (seq space (f x))))))
```
Comment (aalekseyev) [inline src/stdune/pp.ml]: What does this do if `f x` is multiline?
I guess you'd want it do do this:

```
- first line
  next line
- second item
  second line of second item
```

but it doesn't look like the code would do that. (not sure)

File: src/stdune/user_message.mli [hunk 2]
```diff
@@ -0,0 +1,66 @@
+(** A message for the user *)
+
+(** User messages are styled document that can be printed to the
+    console or in the log file. *)
+
+(** Styles that can be used inside messages *)
+module Style : sig
```
Comment (aalekseyev) [inline src/stdune/user_message.mli]: I don't really understand what the abstract styles mean. It looks like most of them are not used?
If we want to make something future-proof, shouldn't it just make it extensible? For example, agree to never remove or rename a style, but that we can add a style later.

File: src/stdune/ansi_color.ml [hunk 3]
```diff
@@ -136,7 +136,20 @@ module Styles = struct
     Style.escape_sequence l
 end
 
-module Render = Pp.Renderer.Make(struct
+let term_supports_color = lazy (
```
Comment (aalekseyev) [inline src/stdune/ansi_color.ml]: It looks like there is very similar code in src/colors.ml. Probably one of the versions should be removed.

File: src/stdune/user_message.ml [hunk 4]
```diff
@@ -0,0 +1,116 @@
+module Style = struct
+  type t =
```
Comment (aalekseyev) [inline src/stdune/user_message.ml]: Oh, I guess there's some relationship between this and the code in src/colors.ml, but it's not clear to me what the relationship is. Is the plan to get rid of src/colors.ml then?

File: src/stdune/pp.ml (hunk 1 above)
Comment (diml) [reply src/stdune/pp.ml]: That's what the `box ~indent:2` is for :)

I added a test for it (and found a bug)

File: src/stdune/user_message.mli (hunk 2 above)
Comment (diml) [reply src/stdune/user_message.mli]: They are meant to be symbolic styles that are converted to concrete ones (such as ansi terminal styles) before rendering to a particular output device. I added a bit more doc.

With the old error reporting API, they are passed inside format strings via `"@{<error>"`.

File: src/stdune/ansi_color.ml (hunk 3 above)
Comment (diml) [reply src/stdune/ansi_color.ml]: Indeed, I removed it

File: src/stdune/user_message.ml (hunk 4 above)
Comment (diml) [reply src/stdune/user_message.ml]: Yep. I updated src/colors.ml to remove a bit of the duplication

File: src/stdune/pp.ml (hunk 1 above)
Comment (aalekseyev) [reply src/stdune/pp.ml]: Ah, sorry, I didn't understand what `indent` was doing. Looks good.

**Pull request merged**